app/actions/views/permalink.ts

```diff
@@ -2,44 +2,113 @@
 
 import {intlShape} from 'react-intl';
-import {Keyboard} from 'react-native';
+import {Alert, Keyboard} from 'react-native';
 
 import {showModalOverCurrentContext} from '@actions/navigation';
 import {loadChannelsByTeamName} from '@actions/views/channel';
+import {getPost} from '@actions/views/post';
 import {selectFocusedPostId} from '@mm-redux/actions/posts';
-import type {DispatchFunc} from '@mm-redux/types/actions';
+import {getChannel, joinChannel} from '@mm-redux/actions/channels';
+import {General} from '@mm-redux/constants';
+import type {DispatchFunc, GetStateFunc} from '@mm-redux/types/actions';
 import {permalinkBadTeam} from '@utils/general';
 import {changeOpacity} from '@utils/theme';
+import {getCurrentUserId} from '@mm-redux/selectors/entities/common';
+import {getCurrentUserRoles} from '@mm-redux/selectors/entities/users';
+import {isSystemAdmin} from '@mm-redux/utils/user_utils';
+import {isChannelMember} from '@utils/channels';
 
 export let showingPermalink = false;
 
 export function showPermalink(intl: typeof intlShape, teamName: string, postId: string, openAsPermalink = true) {
-    return async (dispatch: DispatchFunc) => {
+    return async (dispatch: DispatchFunc, getState: GetStateFunc) => {
         const loadTeam = await dispatch(loadChannelsByTeamName(teamName, permalinkBadTeam.bind(null, intl)));
-
         if (!loadTeam.error) {
             Keyboard.dismiss();
-            dispatch(selectFocusedPostId(postId));
-
-            if (!showingPermalink) {
-                const screen = 'Permalink';
-                const passProps = {
-                    isPermalink: openAsPermalink,
-                    onClose: () => {
-                        dispatch(closePermalink());
-                    },
-                };
-
-                const options = {
-                    layout: {
-                        componentBackgroundColor: changeOpacity('#000', 0.2),
-                    },
-                };
-
-                showingPermalink = true;
-                showModalOverCurrentContext(screen, passProps, options);
+
+            // Warn before joining private channels through permalink
+            const state = getState();
+            const currentUserId = getCurrentUserId(state);
+            const roles = currentUserId ? getCurrentUserRoles(state) : '';
+            if (isSystemAdmin(roles)) {
+                const postData = await dispatch(getPost(postId));
+                if (postData.error) {
+                    return {error: postData.error};
+                }
+                const channelId = postData.data?.channel_id;
+                if (channelId) {
+                    const {data: isMember} = isChannelMember(state, channelId);
+                    if (!isMember) {
+                        const {data: channelData, error: channelError} = await dispatch(getChannel(channelId));
+                        if (channelError) {
+                            return {error: channelError};
```

Comment:
And what about this error?

Reply:
Thanks! Error handling isn't implemented here or the components using this action. Since returns are not handled, I have added the alerts with proper messages.